cli/command/system/inspect.go

```diff
@@ -43,14 +49,30 @@ func NewInspectCommand(dockerCli *command.DockerCli) *cobra.Command {
 }
 
 func runInspect(dockerCli *command.DockerCli, opts inspectOptions) error {
-	var elementSearcher inspect.GetRefFunc
+	var previousType string
+	var refObjects = make(map[string]inspectionResult)
 	switch opts.inspectType {
 	case "", "container", "image", "node", "network", "service", "volume", "task", "plugin":
-		elementSearcher = inspectAll(context.Background(), dockerCli, opts.size, opts.inspectType)
+		for i, ref := range opts.ids {
+			objectType, result, err := inspectAll(context.Background(), dockerCli, ref, opts.size, opts.inspectType)
+			if err != nil {
+				return err
+			}
+
+			if i > 0 && objectType != previousType {
+				return fmt.Errorf("Error: all objects to inspect were expected to be of type %s (same as %s), but %s is of type %s", previousType, opts.ids[0], ref, objectType)
+			}
+			previousType = objectType
+			refObjects[ref] = result
+		}
 	default:
 		return fmt.Errorf("%q is not a valid value for --type", opts.inspectType)
 	}
-	return inspect.Inspect(dockerCli.Out(), opts.ids, opts.format, elementSearcher)
+	getInspectionResult := func(ref string) (interface{}, []byte, error) {
+		obj := refObjects[ref]
+		return obj.v, obj.raw, obj.err
+	}
+	return inspect.Inspect(dockerCli.Out(), opts.ids, opts.format, getInspectionResult)
```

Comment:
Hum yeah, both on this PR and the old code, `/info` is called for each reference, that could be optimized.

> wondering if we could find a "cleaner" solution; the current implementation basically filters/inspects "outside" of the inspect.Inspect() method, and passes the results in.

It's not that big of a deal, it's almost already the case (the definition of the inspect is outside of `inspect` package), so we can do pretty much anything we want 😛